honeybee/door.py

```diff
@@ -0,0 +1,334 @@
+
+    @property
+    def upper_left_vertices(self):
+        """List of vertices starting from the upper-left corner.
+
+        This property should be used when exporting to EnergyPlus / OpenStudio.
+        """
+        return self._geometry.upper_left_counter_clockwise_vertices
+
+    @property
+    def triangulated_mesh3d(self):
+        """A ladybug_geometry Mesh3D of the door geometry composed of triangles.
+
+        In EnergyPlus / OpenStudio workflows, this property is used to subdivide
+        the door when it has more than 4 vertices. This is necessary since
+        EnergyPlus cannot accept sub-faces with more than 4 vertices.
+        """
+        return self._geometry.triangulated_mesh3d
+
+    @property
+    def normal(self):
+        """A ladybug_geometry Vector3D for the direction the door is pointing.
+        """
+        return self._geometry.normal
+
+    @property
+    def center(self):
+        """A ladybug_geometry Point3D for the center of the door.
+
+        Note that this is the center of the bounding rectangle around this geometry
+        and not the area centroid.
+        """
+        return self._geometry.center
+
+    @property
+    def area(self):
+        """The area of the door."""
+        return self._geometry.area
+
+    @property
+    def perimeter(self):
+        """The perimeter of the door."""
+        return self._geometry.perimeter
+
+    @property
+    def boundary_condition(self):
+        """Boundary condition of this door."""
+        return self._boundary_condition
+
+    @boundary_condition.setter
+    def boundary_condition(self, value):
+        assert isinstance(value, (Outdoors, Surface)), \
+            'Door boundary condition must be Outdoors or Surface. ' \
+            'Got {}'.format(type(value))
+        self._boundary_condition = value
+
+    @property
+    def parent(self):
+        """Parent Face if assigned. None if not assigned."""
+        return self._parent
+
+    @property
+    def has_parent(self):
+        """Boolean noting whether this Door has a parent Face."""
+        return self._parent is not None
+
+    def set_adjacency(self, other_door):
+        """Set this door to be adjacent to another (and vice versa).
+
+        Note that this method does not verify whether the other_door geometry is
+        co-planar or compatible with this one so it is recommended that a test
+        be performed before using this method in order to verify these criteria.
+        The Face3D.is_centered_adjacent() or the Face3D.is_geometrically_equivalent()
+        methods are both suitable for this purpose.
+
+        Args:
+            other_aperture: Another Aperture object to be set adjacent to this one.
+        """
+        assert isinstance(other_door, Door), \
+            'Expected Door. Got {}.'.format(type(other_door))
+        self.boundary_condition = Surface(other_door)
+        other_door.boundary_condition = Surface(self)
+
+    def move(self, moving_vec):
```

Comment:
I can see 4 arguments for having this method:
1. There are likely many times where you generate a door/aperture in the center of a face but you want to move it to the side of the face.
2. Having the ability to move a door may help us manage doors that share an edge with their parent (something that is legal in E+ but could be problematic in geometry interfaces like Rhino).
3. Remember that the parent face still has an `are_sub_faces_valid` method that can catch if the door has been moved to an illegal location.
4. We probably at least need a hidden version of this method, which will be called by the parent face when it is moved.

Reply:
Agree that 1 and 2 are useful cases. So maybe to ensure that the door stays in the same plane then we probably should ask for a 2d vector to move the subface geometry inside the parent face plane.

Reply:
Glad that you agree with points 1 and 2. While I know that it sounds good in an abstract sense to use a Vector2D, I can guarantee you that it's not nearly as intuitive as you think and, as a result, it is more prone to errors. This is primarily because there's no guarantee that the plane of a given Face has an X or Y-axis pointing in a certain direction.  For example, when you rotate a Face, you rotate the plane with it and, while this is great for cases like a shoebox model and ensuring that the generated mesh grid is always oriented along the dimensions of the box, it also means that you easily end up with cases where the Face plane has a Y-axis pointing in the negative Z direction, which is the opposite of what you would like for a case like moving a door. I use the rotation as an example but there are many other ways that the axis can change from what you expect and it's a very common condition when geometry is coming from CAD interfaces and we have to generate the plane from the vertices.  At least in 3D, you know what direction is up and which is down, which makes it much more likely that you use the right vector. If you still eel strongly about it, I can add a method that accepts a Vector2D but I am going to insist that the 3D version be kept and I would write all of the warnings that I noted above into the method that uses the Vector2D.

Reply:
I see your points but it should not be as complicated. Since you have methods that sorts the points, etc it should not be hard to align the plane with bounding box of the face if needed but for now let's keep it as is and see if anyone ever complains about it.